From 334cea3a49565f8a75a211d19b275e023903b609 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Fri, 17 May 2013 11:59:30 +0100
Subject: [PATCH 059/333] NetworkControl: Add realtime display of MythEvents

This patch displays all MythEvents to NetworkControl clients.

Syntax:
set events on|off|0|1
query events -> On|Off

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/programs/mythfrontend/networkcontrol.cpp |   45 ++++++++++++++++++++++-
 mythtv/programs/mythfrontend/networkcontrol.h   |    4 ++
 2 files changed, 47 insertions(+), 2 deletions(-)

diff --git a/mythtv/programs/mythfrontend/networkcontrol.cpp b/mythtv/programs/mythfrontend/networkcontrol.cpp
index 2920002..df08e23 100644
--- a/mythtv/programs/mythfrontend/networkcontrol.cpp
+++ b/mythtv/programs/mythfrontend/networkcontrol.cpp
@@ -288,10 +288,11 @@ void NetworkControl::run(void)
 
 void NetworkControl::processNetworkControlCommand(NetworkCommand *nc)
 {
-    QMutexLocker locker(&clientLock);
     QString result;
 
+    clientLock.lock();
     int clientID = clients.indexOf(nc->getClient());
+    clientLock.unlock();
 
     if (is_abbrev("jump", nc->getArg(0)))
         result = processJump(nc);
@@ -387,7 +388,7 @@ void NetworkControl::newConnection(QTcpSocket *client)
     notifyDataAvailable();
 }
 
-NetworkControlClient::NetworkControlClient(QTcpSocket *s)
+NetworkControlClient::NetworkControlClient(QTcpSocket *s) : m_bEvents(false)
 {
     m_socket = s;
     m_textStream = new QTextStream(s);
@@ -1022,6 +1023,10 @@ QString NetworkControl::processQuery(NetworkCommand *nc)
             return QString("ERROR: See 'help %1' for usage information "
                            "(parameters mismatch)").arg(nc->getArg(0));
     }
+    else if (is_abbrev("events", nc->getArg(1)))
+    {
+        result = nc->getClient()->WantsEvents() ? "On" : "Off";
+    }
     else
         return QString("ERROR: See 'help %1' for usage information")
                        .arg(nc->getArg(0));
@@ -1063,6 +1068,21 @@ QString NetworkControl::processSet(NetworkCommand *nc)
 
         return result;
     }
+    else if (is_abbrev("events", nc->getArg(1)))
+    {
+        QString res = "OK";
+
+        if (nc->getArgCount() < 3)
+            res = "ERROR: Missing event value.";
+        else if (nc->getArg(2) == "0" || nc->getArg(2).toLower() == "off")
+            nc->getClient()->WantsEvents(false);
+        else if (nc->getArg(2) == "1" || nc->getArg(2).toLower() == "on")
+            nc->getClient()->WantsEvents(true);
+        else
+            res = "ERROR: Invalid event value.";
+
+        return res;
+    }
 
     return QString("ERROR: See 'help %1' for usage information")
                    .arg(nc->getArg(0));
@@ -1521,6 +1541,27 @@ void NetworkControl::customEvent(QEvent *e)
     {
         MythEvent *me = (MythEvent *)e;
         QString message = me->Message();
+        QString data = (me->ExtraDataCount() > 0 && me->ExtraData(0) != "empty") ?
+            " [" + me->ExtraDataList().join("][") + "]" : QString();
+
+        bool notify = false;
+        clientLock.lock();
+        for ( QList<NetworkControlClient*>::const_iterator it = clients.begin();
+            it != clients.end(); ++it)
+        {
+            NetworkControlClient* ncc = *it;
+            if (ncc->WantsEvents())
+            {
+                nrLock.lock();
+                networkControlReplies.push_back(new NetworkCommand(ncc, message + data));
+                nrLock.unlock();
+                notify = true;
+            }
+        }
+        clientLock.unlock();
+
+        if (notify)
+            notifyDataAvailable();
 
         if (message.startsWith("MUSIC_CONTROL"))
         {
diff --git a/mythtv/programs/mythfrontend/networkcontrol.h b/mythtv/programs/mythfrontend/networkcontrol.h
index 810744c..853a266 100644
--- a/mythtv/programs/mythfrontend/networkcontrol.h
+++ b/mythtv/programs/mythfrontend/networkcontrol.h
@@ -31,6 +31,9 @@ class NetworkControlClient : public QObject
     QTcpSocket  *getSocket()     { return m_socket; }
     QTextStream *getTextStream() { return m_textStream; }
 
+    bool WantsEvents() const { return m_bEvents; }
+    void WantsEvents(bool b) { m_bEvents = b; }
+
   signals:
     void commandReceived(QString&);
 
@@ -40,6 +43,7 @@ class NetworkControlClient : public QObject
   private:
     QTcpSocket  *m_socket;
     QTextStream *m_textStream;
+    bool m_bEvents;
 };
 
 class NetworkCommand : public QObject
-- 
1.7.9.5

